capsules/src/sdcard.rs

```diff
@@ -0,0 +1,1500 @@
+                    self.app_state.map(|appst| { appst.read_buffer = Some(slice); });
+                }
+                ReturnCode::SUCCESS
+            }
+
+            // Pass write buffer in from application
+            1 => {
+                if self.app_state.is_none() {
+                    // create new app state
+                    self.app_state.put(AppState {
+                        callback: None,
+                        read_buffer: None,
+                        write_buffer: Some(slice),
+                    });
+                } else {
+                    // app state exists, set write buffer
+                    self.app_state.map(|appst| { appst.write_buffer = Some(slice); });
+                }
+                ReturnCode::SUCCESS
+            }
+
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    fn subscribe(&self, subscribe_num: usize, callback: Callback) -> ReturnCode {
+        match subscribe_num {
+            // Set callback
+            0 => {
+                if self.app_state.is_none() {
+                    // create new app state
+                    self.app_state.put(AppState {
+                        callback: Some(callback),
+                        read_buffer: None,
+                        write_buffer: None,
+                    });
+                } else {
+                    // app state exists, set callback
+                    self.app_state.map(|appst| { appst.callback = Some(callback); });
+                }
+                ReturnCode::SUCCESS
+            }
+
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    fn command(&self, command_num: usize, data: usize, _: AppId) -> ReturnCode {
+        match command_num {
+            // check if present
+            0 => ReturnCode::SUCCESS,
+
+            // is_installed
+            1 => {
+                let value = self.sdcard.is_installed() as usize;
+                ReturnCode::SuccessWithValue { value: value }
+            }
+
+            // initialize
+            2 => self.sdcard.initialize(),
+
+            // read_block
+            3 => {
+                self.kernel_buf.take().map_or(ReturnCode::EBUSY, |kernel_buf| {
+                    self.sdcard.read_blocks(kernel_buf, data as u32, 1)
+                })
+            }
+
+            // write_block
+            4 => {
+                self.app_state.map_or(ReturnCode::ENOMEM, |app_state| {
+                    app_state.write_buffer.as_mut().map_or(ReturnCode::ENOMEM, |write_buffer| {
+                        self.kernel_buf.take().map_or(ReturnCode::EBUSY, |kernel_buf| {
+                            // copy over write data from application
+                            // Limit to minimum length between kernel_buf,
+                            // write_buffer, and 512 (block size)
+                            for ((kernel_byte, &write_byte), _) in
+                                kernel_buf.iter_mut().zip(write_buffer.iter()).zip(0..512) {
```

Comment:
Why is the final `0..512` needed? Is it just to restrict the size to 512? If so, you can instead apply `take(512)` to either of the other iterators, which returns an iterator with _at most_ 512 elements:

```rust
for (k, w) in kernel_buf.iter_mut().zip(write_buffer.iter()).take(152) {
    *k = *w;
}
```